proof.go

```diff
@@ -141,6 +142,61 @@ func (proof Proof) IsEmptyProof() bool {
 	return proof.start == proof.end && len(proof.nodes) == 0 && len(proof.leafHash) == 0
 }
 
+func (proof Proof) isValidEmptyRangeProof(nth *NmtHasher, nID namespace.ID, root []byte, leaves [][]byte, checkNS bool) bool {
+	if !proof.IsEmptyProof() || len(leaves) != 0 {
+		return false
+	}
+
+	if !checkNS {
+		return true
+	}
+
+	nIDLen := nID.Size()
+	rootMin := namespace.ID(MinNamespace(root, nIDLen))
+	rootMax := namespace.ID(MaxNamespace(root, nIDLen))
+
+	// empty proofs are always rejected unless 1) nID is outside the range of
+	// namespaces covered by the root 2) the root represents an empty tree, since
+	// it purports to cover the zero namespace but does not actually include
+	// any such nodes
+	return nID.Less(rootMin) || rootMax.Less(nID) || bytes.Equal(root, nth.EmptyRoot())
+}
+
+// ComputeAndValidateLeafHashes validates and hashes a list of leaves using the provided NMT hasher.
+func ComputeAndValidateLeafHashes(nth *NmtHasher, nid namespace.ID, leaves [][]byte) ([][]byte, error) {
+	hashes := make([][]byte, len(leaves))
+	for i, d := range leaves {
+		if nth.ValidateLeaf(d) != nil {
+			return nil, fmt.Errorf("leaf hash does not match the NMT hasher's hash format")
```

Comment:
why does this error message contain "hash" when the leaf hasn't been hashed yet. I would've expected

```suggestion
			return nil, fmt.Errorf("leaf does not match the NMT hasher's hash format")
```